aten/src/ATen/function_wrapper.py

```diff
@@ -148,6 +149,15 @@ def TypedDict(name, attrs, total=True):  # type: ignore
     ${initializer}
 }""")
 
+UNIFY_ACTUAL = CodeTemplate("""\
+${actual}.type() != (**new_type) ? (**new_type).copy(${actual}) : ${actual}\
+""")
+UNIFY_TYPES = CodeTemplate("""\
+    if (auto new_type = unifyTypes(${to_unify_actuals})) {
+        return (*new_type)->${api_name}(${unified_actuals});
```

Comment:
why do we dispatch here rather than just continuing on to the broadcast code?

Reply:
Because once you arrive at this method you are in a concrete `at::Type` implementation, which means that it has already been determined what the `s_add` below will dispatch to. This is bad, because e.g. the type of `self` can change because of the unification (or even all arguments can get promoted!), so we need to re-dispatch to the new type.

Reply:
I see.  Looking at it now -- it does seem a little strange to change the type dispatch inside of Type -- why didn't we dispatch to Type "correctly" in the first place?

Typically we put functionality in Type for two reasons (as opposed to functional + method API):
1) it was used as an API itself, so if we didn't put functionality in Type it wouldn't work via the Type API correctly.  This isn't true anymore, at least in Variable we no longer dispatch directly to Type (folks may still be using these APIs though, but they should use the functional or method API instead).
2) Both the functional and method API dispatch to Type, so you only have to do the implementation once.  This still exists.

It does seem nicer to move this functionality to the functional + method API, because it seems like it would work for native functions as well (I'm assuming here that it is easier than implementing it in both non-native and native Type methods, but maybe that's wrong).  And it's nicer to have Type implementations be as simple as possible, because inheriting in Variable is complex.  Unlike broadcasting (which has a number of different implementations based on the function), I can't see a reason why we wouldn't want to have this functionality with native functions -- do you agree?

What do you think @colesbury?

Reply:
BTW I'm happy to get this in (as is) before we figure this out completely -- it would be nice if we could fit broadcasting into this story as well.  For one this would allow us to remove the "s_" magic for derivatives.yaml -- how when I write the derivative for a built-in function `foo` it's actually for `s_foo`, but when I write the derivative for a native function `bar` it's for `bar`, not `s_bar`.  But I don't know how to push the broadcasting into method/functional for native functions.

Reply:
By functional + method API I mean instead of putting the unifyTypes code in Type.cpp, you put it in Functions.cpp (if there's a function) and TensorMethods.h (if there's a method).

So, e.g. at::mul(Tensor self, Tensor other) is going to be something like:
`
return infer_type(self).mul(self, other);
`
and it becomes something like:
`
new_type = ...
return new_type.mul(self.toType(new_type), other.toType(new_type)
`

similarly for methods:
`
return self.type().mul(self, other)
`
becomes something similar to the above.  As I mentioned above, this won't work if you call the Type method directly, but that shouldn't really be viewed as an API, but an implementation detail.

Reply:
Oh I see! I can do that.

So just to clarify: if I want to cover all entry points, then I need to change both `tensor_method_declarations` and `function_declarations`, correct?

Reply:
Would moving the logic out of Type cause issues with tracing?

Reply:
Good question; I'm also not even sure this is the right solution, which is why I said I think it's fine to put this in as is.  What do you think @colesbury?

Reply:
The question I think is what we view Type as being.  Right now, it's:
1) a central place for ops to go so we can implement Variable
2) where broadcast happens
3) the way TH-style dispatch is handled

As written, this complicates the 3) story, but I'm not sure if that's important or not.

Reply:
Why would it be an issue with tracing? We usually want to see all the broadcasts in the IR, and the same applies to type cases. Moving the casts above in the call chain would only increase the chance that they won't be overridden, right? On the other hand, IIRC at the moment we only override at the method/function level so both will be captured and it doesn't matter.

I don't have a strong opinion about where this should live, but since it is very similar to broadcasting I think it makes sense to keep them close. Let me know what are your thoughts, because I'd like to fix and get this in.